modules/tensor_mechanics/src/materials/HyperElasticPhaseFieldIsoDamage.C

```diff
@@ -0,0 +1,155 @@
+  params.addParam<Real>("damage_stiffness", 1e-8, "Avoid zero after complete damage");
+  params.addParam<Real>("zero_tol", 1e-12, "Tolerance for numerical zero");
+  params.addParam<Real>("zero_perturb", 1e-8, "Perturbation value when strain value less than numerical zero");
+  params.addParam<Real>("perturbation_scale_factor", 1e-5, "Perturbation scale factor");
+  params.addRequiredCoupledVar("c", "Damage variable");
+
+  return params;
+}
+
+HyperElasticPhaseFieldIsoDamage::HyperElasticPhaseFieldIsoDamage(const InputParameters & parameters) :
+    FiniteStrainHyperElasticViscoPlastic(parameters),
+    _num_stiffness(getParam<bool>("numerical_stiffness")),
+    _kdamage(getParam<Real>("damage_stiffness")),
+    _zero_tol(getParam<Real>("zero_tol")),
+    _zero_pert(getParam<Real>("zero_perturb")),
+    _pert_val(getParam<Real>("perturbation_scale_factor")),
+    _c(coupledValue("c")),
+    _save_state(false),
+    _G0(declareProperty<Real>(_base_name + "G0")),
+    _dG0_dstrain(declareProperty<RankTwoTensor>(_base_name + "dG0_dstrain")),
+    _dstress_dc(declarePropertyDerivative<RankTwoTensor>(_base_name + "stress", getVar("c", 0)->name())),
+    _etens(LIBMESH_DIM)
+{
+}
+
+void
+HyperElasticPhaseFieldIsoDamage::computePK2StressAndDerivative()
+{
+  computeElasticStrain();
+
+  _save_state = true;
+  computeDamageStress();
+  _pk2[_qp] = _pk2_tmp;
+
+  _save_state = false;
+  if (_num_stiffness)
+    computeNumStiffness();
+
+  if (_num_stiffness)
+    _dpk2_dce = _dpk2_dee * _dee_dce;
+
+  _dce_dfe.zero();
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+      for (unsigned int k = 0; k < LIBMESH_DIM; ++k)
+      {
+        _dce_dfe(i,j,k,i) = _dce_dfe(i,j,k,i) + _fe(k,j);
+        _dce_dfe(i,j,k,j) = _dce_dfe(i,j,k,j) + _fe(k,i);
+      }
+
+  _dpk2_dfe = _dpk2_dce * _dce_dfe;
+}
+
+void
+HyperElasticPhaseFieldIsoDamage::computeDamageStress()
+{
+  Real lambda = _elasticity_tensor[_qp](0, 0, 1, 1);
+  Real mu = _elasticity_tensor[_qp](0, 1, 0, 1);
+
+  Real c = _c[_qp];
+  Real xfac = std::pow(1.0-c, 2.0) + _kdamage;
+
+  std::vector<Real> w;
+  RankTwoTensor evec;
+  _ee.symmetricEigenvaluesEigenvectors(w,evec);
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    _etens[i].vectorOuterProduct(evec.column(i), evec.column(i));
+
+  Real etr = 0.0;
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    etr += w[i];
+
+  Real etrpos=(std::abs(etr)+etr)/2.0;
+  Real etrneg=(std::abs(etr)-etr)/2.0;
+
+  RankTwoTensor pk2pos, pk2neg;
+
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    pk2pos += _etens[i] * ( lambda * etrpos + 2.0 * mu * (std::abs(w[i]) + w[i])/2.0);
+    pk2neg += _etens[i] * ( lambda * etrneg + 2.0 * mu * (std::abs(w[i]) - w[i])/2.0);
+  }
+
+  _pk2_tmp = pk2pos * xfac - pk2neg;
+
+  std::vector<Real> epos(LIBMESH_DIM);
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+    epos[i] = (std::abs(w[i]) + w[i])/2.0;
```

Comment:
This loop and vector variable should get moved into the if block below! If `_save_state` is false this gets computed for nothing.